File: python3/Dynamic Programming/longest_palindrome_in_a_string_expanding.py
from typing import Tuple


def longest_palindrome_in_a_string_expanding(s: str) -> str:
    n = len(s)
    start, max_len = 0, 0
    for center in range(n):
        # Check for odd-length palindromes.
        odd_start, odd_length = expand_palindrome(center, center, s)
        if odd_length > max_len:
            start = odd_start
            max_len = odd_length
        # Check for even-length palindromes.
        if center < n - 1 and s[center] == s[center + 1]:
            even_start, even_length = expand_palindrome(center, center + 1, s)
            if even_length > max_len:
                start = even_start
                max_len = even_length
    return s[start : start + max_len]

# Expands outward from the center of a base case to identify the start 
# index and length of the longest palindrome that extends from this 
# base case.
def expand_palindrome(left: int, right: int, s: str) -> Tuple[int, int]:
    while left > 0 and right < len(s) - 1 and s[left - 1] == s[right + 1]:
        left -= 1
        right += 1
    return left, right - left + 1

File: kotlin/Dynamic Programming/LongestPalindromeInAStringExpanding.kt
fun longestPalindromeInAStringExpanding(s: String): String {
    val n = s.length
    var start = 0
    var maxLen = 0
    for (center in 0 until n) {
        // Check for odd-length palindromes.
        val (oddStart, oddLength) = expandPalindrome(center, center, s)
        if (oddLength > maxLen) {
            start = oddStart
            maxLen = oddLength
        }
        // Check for even-length palindromes.
        if (center < n - 1 && s[center] == s[center + 1]) {
            val (evenStart, evenLength) = expandPalindrome(center, center + 1, s)
            if (evenLength > maxLen) {
                start = evenStart
                maxLen = evenLength
            }
        }
    }
    return s.substring(start, start + maxLen)
}

// Expands outward from the center of a base case to identify the start
// index and length of the longest palindrome that extends from this
// base case.
fun expandPalindrome(left: Int, right: Int, s: String): Pair<Int, Int> {
    var l = left
    var r = right
    while (l > 0 && r < s.length - 1 && s[l - 1] == s[r + 1]) {
        l--
        r++
    }
    return l to r - l + 1
}
